network/stream.go

```diff
@@ -51,41 +53,65 @@ func (s *streamService) handleStream(stream lp2pnetwork.Stream) {
 	s.eventCh <- event
 }
 
-// SendRequest sends a message to a specific peer.
-// If a direct connection can't be established, it attempts to connect via a relay node.
-// Returns an error if the sending process fails.
-func (s *streamService) SendRequest(msg []byte, pid lp2peer.ID) error {
+// SendRequest sends a message to a specific peer, assuming there is already a direct connection.
+//
+// For simplicity, we do not use bi-directional streams.
+// Each time a peer wants to send a message, it creates a new stream.
+//
+// For more details on stream multiplexing, refer to: https://docs.libp2p.io/concepts/multiplex/overview/
+func (s *streamService) SendRequest(msg []byte, pid lp2peer.ID) (lp2pnetwork.Stream, error) {
 	s.logger.Trace("sending stream", "to", pid)
 	_, err := s.host.Peerstore().SupportsProtocols(pid, s.protocolID)
 	if err != nil {
-		return LibP2PError{Err: err}
+		return nil, LibP2PError{Err: err}
 	}
 
 	// To prevent a broken stream from being open forever.
-	ctxWithTimeout, cancel := context.WithTimeout(s.ctx, 20*time.Second)
+	ctxWithTimeout, cancel := context.WithTimeout(s.ctx, 5*time.Second)
 	defer cancel()
 
-	// Attempt to open a new stream to the target peer assuming there's already direct a connection
+	// Attempt to open a new stream to the peer, assuming there's already a direct connection.
 	stream, err := s.host.NewStream(
 		lp2pnetwork.WithNoDial(ctxWithTimeout, "should already have connection"), pid, s.protocolID)
 	if err != nil {
-		return LibP2PError{Err: err}
+		return nil, LibP2PError{Err: err}
 	}
 
-	deadline, _ := ctxWithTimeout.Deadline()
-	_ = stream.SetDeadline(deadline)
-
 	_, err = stream.Write(msg)
 	if err != nil {
 		_ = stream.Reset()
 
-		return LibP2PError{Err: err}
+		return nil, LibP2PError{Err: err}
 	}
 
 	err = stream.CloseWrite()
 	if err != nil {
-		return LibP2PError{Err: err}
+		return nil, LibP2PError{Err: err}
 	}
 
-	return nil
+	// We need to close the stream once it is read by the receiver.
+	// If, for any reason, the receiver doesn't close the stream, we need to close it after a timeout.
+	go func() {
+		timer := time.NewTimer(s.timeout)
+		closed := make(chan bool)
+
+		go func() {
+			// We need only one byte to read the EOF.
+			buf := make([]byte, 1)
+			_, _ = stream.Read(buf)
+			closed <- true
+		}()
+
+		select {
+		case <-timer.C:
+			s.logger.Warn("stream timeout", "to", pid)
+			_ = stream.Close()
+
```

Comment:
Please close channel `close(closed)` here

Reply:
No need. The garbage collector handles channels that are no longer needed.